doc: kconfig: Move misc. scattered Kconfig docs to a new Kconfig section in User and Developer Guides

ulfalizer opened the pull request:
A bunch of important Kconfig documentation was scattered throughout different pages and hard to find. This PR creates a new *Kconfig* section in *User and Developer Guides*, and moves stuff into it.

`doc/guides/kconfig/index.rst` now looks like this (previously, it was the *Kconfig - Tips and Best Practices* page):

```rst
Kconfig
#######

.. toctree::
   :maxdepth: 1

   menuconfig.rst
   setting.rst
   tips.rst
   preprocessor-functions.rst
   extensions.rst
```

A bunch of minor documentation tweaks and cleanups are included as well.

Commits:

```
doc: kconfig: Touch up preprocessor docs and put it on a separate page

Touch up the Kconfig preprocessor function documentation a bit and add
some more details to the example. Use the Kconfig preprocessor call
syntax instead of Python syntax in the function synopses.

Also move the preprocessor docs to a separate page, and add a top-level
Kconfig section to User and Developer Guides that links to the Kconfig
tips page and the preprocessor page.

Trying to avoid the Kconfig tips page becoming a dumping ground for
random Kconfig documentation.
```

```
doc: kconfig: Move extension docs from porting guide to user guide

The Kconfiglib Kconfig extensions are documented in the porting guide,
which is a pretty weird place to document them. Document them under a
new page in the Kconfig section of User and Developer Guides instead.

Also remove the section about the old Zephyr Kconfig behavior for
defaults (tried last-to-first instead of first-to-last). The behavior
was changed 16 months ago (and was undocumented before that), so it
might not be that useful to mention anymore.

Piggyback misc. small language cleanups and organization nits.
```

```
doc: kconfig: Move menuconfig/guiconfig docs to user guide

Move the documentation for the menuconfig and guiconfig interfaces from
Application Development to a new Interactive Kconfig Interfaces page in
User and Developer Guides.

Also tweak Application Development to mention the configuration
interfaces earlier. It's good to encourage people to try things out in
menuconfig, because dependencies often get lost when people seldom run
it and just hand-edit files.
```

```
doc: application: Remove accidental backslashes in Kconfig sample

The SOC Definitions section had backslashes before * in globbing
'source's.

Also change the language for the code block from 'console' to 'none'.
Kconfig isn't shell.
```

```
doc: kconfig: Put all documentation on setting symbols on a new page

Add a new 'Setting Kconfig configuration values' page in the new Kconfig
section in 'User and Developer Guides'. Move all the information on
setting Kconfig symbols from 'Application Development' and the board
porting guide into it. The same page now covers both configuration files
and Kconfig.defconfig files.

Add links to the new page in various places to make it easy to find.
Also add some more references to the top-level Kconfig page and other
Kconfig pages.

A lot of stuff was rewritten while moving it over (the CONF_FILE
documentation has been cleaned up in particular). Some new information
has been added as well, like a tip re. minimal configurations being
helpful when making Kconfig settings permanent, and a warning re.
dependencies being ORed rather than ANDed when defining a symbol in
multiple locations.

Fixes: #20915
```

```
doc: kconfig: Add Kconfig intro to index page 

Put a short Kconfig overview on the index page that links to the other
Kconfig documentation.
```

ulfalizer committed:
doc: kconfig: Touch up preprocessor docs and put it on a separate page

Touch up the Kconfig preprocessor function documentation a bit and add
some more details to the example. Use the Kconfig preprocessor call
syntax instead of Python syntax in the function synopses.

Also move the preprocessor docs to a separate page, and add a top-level
Kconfig section to User and Developer Guides that links to the Kconfig
tips page and the preprocessor page.

Trying to avoid the Kconfig tips page becoming a dumping ground for
random Kconfig documentation.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
doc: kconfig: Move extension docs from porting guide to user guide

The Kconfiglib Kconfig extensions are documented in the porting guide,
which is a pretty weird place to document them. Document them under a
new page in the Kconfig section of User and Developer Guides instead.

Also remove the section about the old Zephyr Kconfig behavior for
defaults (tried last-to-first instead of first-to-last). The behavior
was changed 16 months ago (and was undocumented before that), so it
might not be that useful to mention anymore.

Piggyback misc. small language cleanups and organization nits.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
doc: kconfig: Move menuconfig/guiconfig docs to user guide

Move the documentation for the menuconfig and guiconfig interfaces from
Application Development to a new Interactive Kconfig Interfaces page in
User and Developer Guides.

Also tweak Application Development to mention the configuration
interfaces earlier. It's good to encourage people to try things out in
menuconfig, because dependencies often get lost when people seldom run
it and just hand-edit files.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer commented:
@erwango 
Will move some more stuff from the board porting guide to the new Kconfig section later as well, fixing https://github.com/zephyrproject-rtos/zephyr/issues/20915. Splitting the changes up a bit to be more manageable.

ulfalizer committed:
doc: kconfig: Put all documentation on setting symbols on a new page

Add a new 'Setting Kconfig configuration values' page in the new Kconfig
section in 'User and Developer Guides'. Move all the information on
setting Kconfig symbols from 'Application Development' and the board
porting guide into it. The same page now covers both configuration files
and Kconfig.defconfig files.

Add links to the new page in various places to make it easy to find.
Also add some more references to the top-level Kconfig page and other
Kconfig pages.

A lot of stuff was rewritten while moving it over (the CONF_FILE
documentation has been cleaned up in particular). Some new information
has been added as well, like a tip re. minimal configurations being
helpful when making Kconfig settings permanent, and a warning re.
dependencies being ORed rather than ANDed when defining a symbol in
multiple locations.

Fixes: #20915

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
doc: application: Remove accidental backslashes in Kconfig sample

The SOC Definitions section had backslashes before * in globbing
'source's.

Also change the language for the code block from 'console' to 'none'.
Kconfig isn't shell.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer commented:
> @erwango
> Will move some more stuff from the board porting guide to the new _Kconfig_ section later as well, fixing #20915. Splitting the changes up a bit to be more manageable.

Done now. See the latest commit ("doc: kconfig: Put all documentation on setting symbols on a new page"). Lots of rewriting and some new documentation there as well.

Pretty much complete documentation for Kconfig can now be found in the new Kconfig section under 'User and Developer' guides. Phew...

ulfalizer commented:
If you want to review the complete changes, read through the pages in the new Kconfig section in User and Developer Guides, and check that the links from Application Development and the board porting guide look fine.

erwango reviewed (approved):
Fine with me from board configuration perspective

carlescufi commented on doc/application/index.rst:
```diff
@@ -1178,316 +1185,10 @@ documentation `runningcmake`_ .
-following ways. The simplest option is to just have a single :file:`prj.conf`
-file.
-
-1. If :makevar:`CONF_FILE` is set in :file:`CMakeLists.txt` (**before including
```
You have removed a bunch of text from these sections, not sure if *too* much. I found some of these tips to be very useful. Also, should we keep a minimal description of Kconfig and its usage in this .rst file? @mbolivar @SebastianBoe

ulfalizer replied:
What tips are you thinking of? Note that a bunch of stuff has just been rephrased to be shorter or moved around.

mbolivar replied:
> Also, should we keep a minimal description of Kconfig and its usage in this .rst file? @mbolivar @SebastianBoe

I did not read the diff in full, but the existing pointers to the new doc that were left in the application guide seem OK to me.

ulfalizer replied:
Added back a short Kconfig section to the Application page that just mentions `prj.conf` and links off to the Kconfig docs, and linked that one instead from the other Application sections. Should make it harder to miss.

ulfalizer replied:
@carlescufi ping

carlescufi reviewed (changes requested):
In general looks good, but I'd like to double-check the changes to the application/index.rst file because in my opinion they might go a bit too far.

dbkinder commented on doc/guides/kconfig/index.rst:
```diff
@@ -1,918 +1,15 @@
-.. contents::
-    :local:
-    :depth: 2
+.. toctree::
```
We could use a short introduction to configuration and Kconfig here, before sending folks off into the detailed topics listed.  Something like this perhaps:
```
There are many different configurations possible for the Zephyr
kernel and subsystems to adapt them for specific application and
platform needs. Many configuration choices can be made without
actually changing the source code, by first configuring the kernel
and subsystem options and then building the source with those options.

Zephyr uses the same Kconfig system used by the Linux Kernel.
Kconfig files (organized in a tree structure within the Zephyr code tree)
describe build-time configuration options called *symbols*, how they’re
grouped into menus and sub-menus, and dependencies between
them that determine what configurations are valid.

The following sections go into details about Kconfig use within the
Zephyr project, how to specify configuration options, tips and
notes about Kconfig, and more.
```

erwango replied:
+1, I think we can even emphasise that this is one of the asset of zephyr to achieve low footprint goal

ulfalizer replied:
Added a similar intro now, along with some misc. info like the header name. Check that it looks okay.

dbkinder reviewed (changes requested):
This looks like a nice organizational improvement, thanks.  One request for a quick intro, otherwise LGTM,

dbkinder commented:
BTW, you can see what this new doc set looks like in the build artifacts: https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/21133/docs/guides/kconfig/index.html

ulfalizer committed:
doc: kconfig: Add Kconfig intro to index page

Put a short Kconfig overview on the index page that links to the other
Kconfig documentation.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

dbkinder commented on doc/guides/kconfig/index.rst:
```diff
@@ -1,918 +1,33 @@
 .. SPDX-License-Identifier: Apache-2.0
 
```
Don’t need the SPDX license tag on doc files.

ulfalizer replied:
Removed all the license tags. Thought it was required to please CI.

dbkinder reviewed (approved):
Other than the stray license tag, docs LGTM

carlescufi reviewed (approved):
LGTM now, thanks!
